ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/FollowerPromptness.hs

```diff
@@ -0,0 +1,214 @@
+{-# LANGUAGE BlockArguments      #-}
+{-# LANGUAGE DerivingStrategies  #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
+{-# LANGUAGE ViewPatterns        #-}
+
+-- | In the context of diffusion pipelining, it is important that (tentative)
+-- followers promptly emit an instruction to roll forward after the tentative
+-- header got set. We ensure this behavior by comparing the timestamps when a
+-- tentative header got set and when a tentative follower emits an instruction
+-- containing it.
+module Test.Ouroboros.Storage.ChainDB.FollowerPromptness (tests) where
+
+import           Control.Monad (forever)
+import           Control.Monad.IOSim (runSimOrThrow)
+import           Control.Tracer (Tracer (..), contramapM, traceWith)
+import           Data.Foldable (for_)
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Set (Set)
+import qualified Data.Set as Set
+
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+import qualified Ouroboros.Network.MockChain.Chain as Chain
+
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Consensus.Storage.ChainDB.API (ChainDB)
+import qualified Ouroboros.Consensus.Storage.ChainDB.API as ChainDB
+import qualified Ouroboros.Consensus.Storage.ChainDB.API.Types.InvalidBlockPunishment as Punishment
+import           Ouroboros.Consensus.Storage.ChainDB.Impl (ChainDbArgs (..))
+import qualified Ouroboros.Consensus.Storage.ChainDB.Impl as ChainDBImpl
+import           Ouroboros.Consensus.Util.Condense (Condense (..))
+import           Ouroboros.Consensus.Util.Enclose
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import           Test.Util.ChainDB
+import           Test.Util.ChainUpdates
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock
+import           Test.Util.Tracer (recordingTracerTVar)
+
+tests :: TestTree
+tests = testGroup "FollowerPromptness"
+    [ testProperty "followerPromptness" prop_followerPromptness
+    ]
+
+prop_followerPromptness :: FollowerPromptnessTestSetup -> Property
+prop_followerPromptness fpts =
+    counterexample ("Trace:\n" <> unlines (ppTrace <$> traceByTime)) $
```

Comment:
Another personal style: I find it more difficult to read long applications when the operators are on the ends of the line _and the operators vary per line_ (`$` or `.&&.`) here.

For whatever reason, we do seem to favor end-of-line-`$` for QuickCheck function applications, but maybe we fall back to the usual leading-operator style for the `.&&.`? So:

```
    counterexample (condense fpts) $
    counterexample ("Instruction timings: " <> condense followerInstrTimings) $
           counterexample ("Failed to pipeline: " <> condense notPipelined) (null notPipelined)
      .&&. counterexample ("Not processed: " <> condense unprocessed)       (null unprocessed)

Reply:
We use trailing `$` and `.&&.` e.g. here:
https://github.com/input-output-hk/ouroboros-network/blob/edbf4ea89ccac7195bd606407d1239b3e7fecab6/ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs#L95-L129
but not so close together, which is indeed confusing. I applied your suggestion of using leading `.&&.` :+1: